api/session_join.go

```diff
@@ -83,11 +84,22 @@ func sessionJoinPost(sh *service.Handler) func(state state.State, r *http.Reques
 
 // validateIntent validates the given join intent.
 // It checks whether or not the peer is missing any of our services and returns an error if one is missing.
-func validateIntent(sh *service.Handler, intent types.SessionJoinPost) error {
+// Also compares each service's daemon version between the joiner and initiator.
+func validateIntent(ctx context.Context, sh *service.Handler, intent types.SessionJoinPost) error {
 	// Reject any peers that are missing our services.
-	for service := range sh.Services {
-		if !shared.ValueInSlice(service, intent.Services) {
-			return fmt.Errorf("Rejecting peer %q due to missing services (%s)", intent.Name, string(service))
+	for _, service := range sh.Services {
+		intentVersion, ok := intent.Services[service.Type()]
+		if !ok {
+			return fmt.Errorf("Rejecting peer %q due to missing services (%s)", intent.Name, string(service.Type()))
+		}
+
+		version, err := service.GetVersion(ctx)
+		if err != nil {
+			return fmt.Errorf("Unable to determine initiator's %s version", service.Type())
```

Comment:
`"Unable to determine initiator's %s version: %w"`

Reply:
Fixed.